src/BlockTypes/SingleProduct.php

```diff
@@ -0,0 +1,140 @@
+	}
+
+
+	/**
+	 * Render the Single Product block
+	 *
+	 * @param array    $attributes Block attributes.
+	 * @param string   $content Block content.
+	 * @param WP_Block $block Block instance.
+	 *
+	 * @return string Rendered block output.
+	 */
+	protected function render( $attributes, $content, $block ) {
+		$classname = $attributes['className'] ?? '';
+
+		$html = sprintf(
+			'<div class="wp-block-woocommerce-single-product %1$s">
+				%2$s
+			</div>',
+			esc_attr( $classname ),
+			$content
+		);
+
+		return $html;
+	}
+
+	/**
+	 * Update the context by injecting the correct post data
+	 * for each one of the Single Product inner blocks.
+	 *
+	 * @param array    $context Block context.
+	 * @param array    $block Block attributes.
+	 * @param WP_Block $parent_block Block instance.
+	 *
+	 * @return array Updated block context.
+	 */
+	public function update_context( $context, $block, $parent_block ) {
+		if ( 'woocommerce/single-product' == $block['blockName']
+			&& isset( $block['attrs']['productId'] ) ) {
+				$this->product_id = $block['attrs']['productId'];
+
+				$this->single_product_inner_blocks_names = array_reverse(
+					$this->extract_single_product_inner_block_names( $block )
+				);
+		}
+
+		$this->replace_post_for_single_product_inner_block( $block, $context );
+
+		return $context;
+	}
+
+	/**
+	 * Extract the inner block names for the Single Product block. This way it's possible
+	 * to map all the inner blocks for a Single Product block and manipulate the data as needed.
+	 *
+	 * @param array $block The Single Product block or its inner blocks.
+	 * @param array $result Array of inner block names.
+	 *
+	 * @return array Array containing all the inner block names of a Single Product block.
+	 */
+	protected function extract_single_product_inner_block_names( $block, &$result = [] ) {
+		if ( isset( $block['blockName'] ) ) {
+			$result[] = $block['blockName'];
+		}
+
+		if ( isset( $block['innerBlocks'] ) ) {
+			foreach ( $block['innerBlocks'] as $inner_block ) {
+				$this->extract_single_product_inner_block_names( $inner_block, $result );
+			}
+		}
+		return $result;
+	}
+
+	/**
+	 * Replace the global post for the Single Product inner blocks and reset it after.
+	 *
+	 * This is needed because some of the inner blocks may use the global post
+	 * instead of fetching the product through the `productId` attribute, so even if the
+	 * `productId` is passed to the inner block, it will still use the global post.
+	 *
+	 * @param array $block Block attributes.
+	 * @param array $context Block context.
+	 */
+	protected function replace_post_for_single_product_inner_block( $block, &$context ) {
+		if ( $this->single_product_inner_blocks_names ) {
+			$block_name = array_pop( $this->single_product_inner_blocks_names );
+
+			if ( $block_name === $block['blockName'] ) {
+				global $post;
+				// phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+				$post = get_post( $this->product_id );
```

Comment:
I would encourage finding an alternative approach here rather than overriding the WordPress `$post` global variable for a few reasons:

* Overwriting WordPress global variables can cause conflicts and unexpected behavior.
* When we overwrite global variables, there's a risk of breaking core functionality or interfering with the functionality of plugins and themes that rely on those. 
* Additionally, it can make the code harder to maintain and debug, as it can be challenging to track down where a particular value is being set or changed.

As an alternative, how about updating the `render` method for each one of the relevant blocks, so they have different behaviors depending on the context? Assuming the `$product_id` value for the block can be fetched from the DB:

example for the Add to cart form:

```php
ob_start();

if ( is_product() ) {
	while ( have_posts() ) {
		the_post();
		global $product;
		do_action( 'woocommerce_' . $product->get_type() . '_add_to_cart' );
	}
} else {
	$product = wc_get_product( $product_id );

	if ( ! $product ) {
		return;
	}

	do_action( 'woocommerce_' . $product->get_type() . '_add_to_cart' );
}

$product = ob_get_clean();
```

What do you think?

Reply:
I agree that this is not the best approach and thank you for suggesting a new one; however, I think this could be handled in a separate PR that would refactor every block to use the approach that we decide on [Using the Blockified Single Product Template](https://github.com/woocommerce/woocommerce-blocks/discussions/8792). Editing the `render` method would not work for the variations blocks, because we cannot edit their render method :/ (ref: `per0F9-qM-p2`)

Reply:
Thank you for bringing this solution here, yes I agree with you we can implement your suggestion and it would be much better. I'll be changing the code to match what you suggested and put a conditional to only apply the global post when it is the post excerpt block, at least until GB approves the change to its core, or we have a dedicated version of the Post Excerpt

Reply:
@nefeline I replaced `replace_post_for_single_product_inner_block` with your suggested solution, but it gives me an error when I go to the frontend, am I missing something?

<img width="1227" alt="image" src="https://user-images.githubusercontent.com/20469356/228928783-2c6d29ef-dbb0-40bc-88a5-e50e3f0655f9.png">

Reply:
Sorry about that! I forgot to share a few changes I did to the AddToCartForm as well to support this change, they are [available here](https://github.com/woocommerce/woocommerce-blocks/compare/feat/add-single-product-block-server-side...try/global-variable-overwrite?expand=1).

Reply:
Thank you for sharing this! Now the error is gone, but I noticed that if I add the block inside the Single Product template and visit a product page, the product that is displayed is not the one that was selected in the Editor, is the product that the user is currently visiting

Reply:
> I noticed that if I add the block inside the Single Product template and visit a product page, the product that is displayed is not the one that was selected in the Editor, is the product that the user is currently visiting

Nice catch! That's because of [this line](https://github.com/woocommerce/woocommerce-blocks/blob/576cc4151bf1587453f9f86d187768debb3ab2b5/src/BlockTypes/SingleProduct.php#L127): if we ditch the `! is_product()` validation we are all set: in reality, it doesn't matter if we are in the context of a single product template or not; we will always want to fetch the content from the block context.